examples/dtls-echo/dtls-client.c

```diff
@@ -210,325 +199,360 @@ static int peer_get_ecdsa_key(struct dtls_context_t *ctx,
-    dst->port =  (unsigned short) DEFAULT_PORT;
+    dtls_connected = 0;
 
-    if (ipv6_addr_from_str(&dst->addr, addr_str) == NULL) {
-        puts("ERROR: init_dtls was unable to load the IPv6 addresses!\n");
-        dtls_context = NULL;
-        return;
+#ifdef TINYDTLS_LOG_LVL
+    dtls_set_log_level(TINYDTLS_LOG_LVL);
+#endif
+
+    /* First, we prepare the UDP Sock */
+    local->port = (unsigned short) CLIENT_PORT;
+    remote->port = (unsigned short) DTLS_DEFAULT_PORT;
+
+    /* Parsing <address>[:<iface>]:Port */
+    int iface = ipv6_addr_split_iface(addr_str);
+    if (iface == -1) {
+        if (gnrc_netif_numof() == 1) {
+            /* assign the single interface found in gnrc_netif_numof() */
+            dst->ifindex = (uint16_t)gnrc_netif_iter(NULL)->pid;
+            remote->netif = (uint16_t)gnrc_netif_iter(NULL)->pid;
+        }
+        else {
+            /* FIXME This probably is not valid with multiple interfaces */
+            dst->ifindex = remote->netif;
+        }
+    }
+    else {
+        if (gnrc_netif_get_by_pid(iface) == NULL) {
+            puts("ERROR: interface not valid");
+            return new_context;
+        }
+        dst->ifindex = (uint16_t)gnrc_netif_iter(NULL)->pid;
+        remote->netif = (uint16_t)gnrc_netif_iter(NULL)->pid;
     }
 
-    ipv6_addr_t addr_dbg;
-    ipv6_addr_from_str(&addr_dbg, addr_str);
+    if (ipv6_addr_from_str((ipv6_addr_t *)remote->addr.ipv6, addr_str) == NULL) {
+        puts("ERROR: unable to parse destination address");
+        return new_context;
+    }
 
-    /*akin to syslog: EMERG, ALERT, CRITC, NOTICE, INFO, DEBUG */
-    dtls_set_log_level(DTLS_LOG_NOTICE);
+    /* Second: We prepare the DTLS Session by means of ctx->app */
+    dst->size = sizeof(uint8_t) * 16 + sizeof(unsigned short);
+    dst->port = remote->port;
 
-    dtls_context = dtls_new_context(addr_str);
-    if (dtls_context) {
-        dtls_set_handler(dtls_context, &cb);
+    /* NOTE: remote.addr.ipv6 and dst->addr are different structures. */
+    if (ipv6_addr_from_str(&dst->addr, addr_str) == NULL) {
+        puts("ERROR: init_dtls was unable to load the IPv6 addresses!");
+        return new_context;
     }
 
-    return;
+    new_context = dtls_new_context(sock);
+    if (new_context) {
+        dtls_set_handler(new_context, &cb);
+    }
+
+    return new_context;
 }
 
-/**
- * This is the "client" part of this program.
- * Will be called each time a message is transmitted.
- */
-static void client_send(char *addr_str, char *data, unsigned int delay)
+static void client_send(char *addr_str, char *data,
+                        unsigned int delay)
 {
-    int8_t iWatch;
     static session_t dst;
-    static int connected = 0;
-    msg_t msg;
+    dtls_context_t *dtls_context = NULL;
+
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+    sock_udp_ep_t remote = SOCK_IPV6_EP_ANY;
+    sock_udp_t sock;
+
+    uint8_t packet_rcvd[DTLS_MAX_BUF];
+    int8_t watch = MAX_TIMES_TRY_TO_SEND;
```

Comment:
Maybe use an unsigned value here.

Reply:
Maybe I'm being too paranoid, but I intentionally used int8_t to avoid an overflow with `(watch > 0)` and because I'm not expecting a value higher than 20 for this variable. Yet, if this is not a good practice, I'll make the suggested change.